replication/driver/replication_driver.go

```diff
@@ -0,0 +1,305 @@
+package driver
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/zrepl/zrepl/replication/report"
+	"github.com/zrepl/zrepl/util/chainlock"
+)
+
+type run struct {
+	l *chainlock.L
+
+	startedAt, finishedAt time.Time
+
+	// the attempts attempted so far:
+	// All but the last in this slice must have finished with some errors.
+	// The last attempt may not be finished and may not have errors.
+	attempts []*attempt
+}
+
+type Planner interface {
+	Plan(context.Context) ([]FS, error)
+}
+
+// an attempt represents a single planning & execution of fs replications
+type attempt struct {
+	planner Planner
+
+	l *chainlock.L
+
+	startedAt, finishedAt time.Time
+	// after Planner.Plan was called, planErr and fss are mutually exclusive with regards to nil-ness
+	// if both are nil, it must be assumed that Planner.Plan is active
+	planErr *timedError
+	fss     []*fs
+}
+
+type timedError struct {
+	Err  error
+	Time time.Time
+}
+
+func newTimedError(err error, t time.Time) *timedError {
+	if err == nil {
+		panic("error must be non-nil")
+	}
+	if t.IsZero() {
+		panic("t must be non-zero")
+	}
+	return &timedError{err, t}
+}
+
+func (e *timedError) IntoReportError() *report.TimedError {
+	if e == nil {
+		return nil
+	}
+	return report.NewTimedError(e.Err.Error(), e.Time)
+}
+
+type FS interface {
+	PlanFS(context.Context) ([]Step, error)
+	ReportInfo() *report.FilesystemInfo
+}
+
+type Step interface {
+	TargetDate() time.Time
+	Step(context.Context) error
+	ReportInfo() *report.StepInfo
+}
+
+type fs struct {
+	fs FS
+
+	l *chainlock.L
+
+	planning struct {
+		done bool
+		err  *timedError
+	}
+
+	// valid iff planning.done && planning.err == nil
+	planned struct {
+		// valid iff planning.done && planning.err == nil
+		stepErr *timedError
+		// all steps, in the order in which they must be completed
+		steps []*step
+		// index into steps, pointing at the step that is currently executing
+		// if step >= len(steps), no more work needs to be done
+		step int
+	}
+}
+
+type step struct {
+	l    *chainlock.L
+	step Step
+}
+
+type ReportFunc func() *report.Report
+type WaitFunc func(block bool) (done bool)
+
+func Do(ctx context.Context, planner Planner) (ReportFunc, WaitFunc) {
+	l := chainlock.New()
+	run := &run{
+		l:         l,
+		startedAt: time.Now(),
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		run.l.Lock()
+		a1 := &attempt{
+			l:         l,
+			startedAt: time.Now(),
+			planner:   planner,
+		}
+		run.attempts = append(run.attempts, a1)
+		run.l.Unlock()
+		a1.do(ctx)
```

Comment:
Error handling is still missing here, given that we let gRPC fail early now:

* build a report (need to drop lock!)
* scan through it, find FSes with errors
* if errors, classify them, determine whether we should wait
* if we wait (non-permanent errors), make sure this is represented in the report
  * state? just a non-zero-valued 'sleep until' value?
* sleep
* build a new attempt
* copy FSes without errors (just pointers) + permanent (non-retryable) errors
* those that we want to retry: new, empty FSes
  * add a predecessor-pointer to the fs struct, need to pass that to the logic package...
* run driver as is, the fact that some FSes are pre-failed shouldn't be a problem
  * write tests for that